### DISCUSSION TITLE
offlineExchange: `teardown` operations are called immediately on the page reload even if a `cache-and-network` call is pending

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#3234] [Q&A]
After the big work on https://github.com/urql-graphql/urql/pull/3200 there is an issue in how `teardown` operations are called on the page reload (when there are cached data).

REPRODUCTION: https://codesandbox.io/p/sandbox/issue-urql-6-duplicated-requests-forked-0ybcue.

### Steps:

1. click on "Todos list"

1. open the browser Dev Tools's console

1. reload the page

1. see in the console `before, else` followed immediately (and not after the fake delay of the backend) `before, op.kind === teardown`

### What I expect

I expect `teardown` to be called after the backend fake delay response.

### More context

I'm using a custom exchange to track pending requests globally:

```ts
const pendingRequestsExchange: Exchange = ({ forward }) => {
  function before(op: Operation) {
    if (op.kind === 'subscription') return;
    
    if (op.kind === 'teardown') {
      console.log('before, op.kind === teardown');
      // decrease the global pending requests here
    } else {
      console.log('before, else');
      // increase the global pending requests here
    }
  }
  function after(op: OperationResult) {
    console.log('after()');
    // decrease the global pending requests here
  }
  return (ops$) => {
    const forward$ = pipe(ops$, tap(before));
    return pipe(forward(forward$), tap(after));
  };
};
```

@kitten [said](https://github.com/urql-graphql/urql/pull/3200#issuecomment-1569872363) that maybe the issue is my custom exchange and that

> you will have to adjust your after call to be more precise. If this is just to keep track of global loading states, I'd suggest to be more accurate with cache-only operations in after and/or maybe adding a delay or debounce here.

but the thing is the `after()` function of the custom exchange is **NEVER** called on the first load.

I do not understand why the `teardown` operation is issued. If this is a `cache-and-network` call it should wait for the server answer, nope?

__Thanks for your work, it's not something I take for granted. Really, thanks.__

### Gif

![issue_teardown](https://github.com/urql-graphql/urql/assets/41120635/25b5c679-b42d-4df4-9be1-7a7ed7c7a526)

**Comment by kitten** [comment]
This is not a bug, but required for us to issue a new operation: https://github.com/urql-graphql/urql/blob/c356285e050c46022192067151c1ec267424062d/exchanges/graphcache/src/offlineExchange.ts#L160

**Reply by frederikhors** [reply]
But then from now on **it will no longer be possible to understand if an operation is in progress**?

**Reply by kitten** [reply]
Yes, of course it will, otherwise many exchanges wouldn't work properly. You, as an exchange author, don't control what kind of signals come in.

If you see a teardown, the operation is becoming inactive,
if you see a regular operation, the operation must be started,
if you see a result, then that's one response,
if you see `hasNext: true` on the result, there'll be more results, until one with `hasNext: false` is delivered.

In short, if you see a `teardown` and no more operations, your operation is completed.
Since this issues `cache-first`, if there's a cache hit then no more operations will be issued unless you issue them

**Reply by frederikhors** [reply]
@kitten, my issue is well described on this issue:

If I reload the page:

1. an ongoing (`cache-and-network`) operation is running

1. `graphcache` sends two signals for this operation, one **IMMEDIATELY** after the other:

    1. operation: `kind: query`

    1. operation: `kind: teardown`

**BUT THE OPERATION IS STILL RUNNING!**

See the gif I posted up here, please.

**Reply by frederikhors** [reply]
I just tried with `@urql/exchange-graphcache": "6.1.1"` and the issue is still there.

**Reply by frederikhors** [reply]
I can't figure out how to fix it unless you or someone else helps me. Please.

I always seem to be intrusive when I post an issue on this project. But I think I always got it right and reported REAL problems. (There are companies that pay to have this.)

Even now the problem is real.

The global pending requests indicator (which worked before these last updates) now, even though there are pending requests, urql gives the wrong signal: the `teardown` (as demonstrated in the REPL).

I open the bug and it gets closed without even having read what I wrote (there is also a gif! Very few people in the world create issues with animated gifs!).

I don't like wasting my time or wasting anyone else's time, but what I'm trying to accomplish is for other people who have asked for the Global Pending Requests Indicator and I'm reporting a NEW problem.

**Can I at least ask you to reopen the bug and treat it as such?**

I'm asking you because otherwise there's no answer to this chat and I'll be waiting endlessly.

Please help, **this is not fixable without your intervention**.

Thank you.

**Reply by kitten** [reply]
You've just simply got to show some patience & restraint, and I only ask of you to debug some problems further yourself.

In this case, I checked your example and it's doing the right thing. After `teardown` is received no more result is dispatched by the `fetchExchange`. This of course happens _after_ the `offlineExchange` has rehydrated, since this re-triggers teardowns and optionally requests.

However, I found that the original network request isn't cancelled when it stalls after the initial request succeeds. This is likely down to when you delay your response, i.e. after headers are sent by your API. I can see that this is confusing and I can put back the old cancellation logic.

**Reply by frederikhors** [reply]
@kitten, unfortunately nor `@urql/core@4.0.9` nor `@urql/core@4.0.10` fix this.

Even worse: now the call is aborted and does not update the data at all!

REPL: https://codesandbox.io/p/sandbox/issue-urql-6-duplicated-requests-forked-619ulm

![issue_teardown](https://github.com/urql-graphql/urql/assets/41120635/371e8194-0562-4cca-9b28-ce814d772745)

**This is another issue now!** :(

> I only ask of you to debug some problems further yourself

I did and I created this issue.

> I checked your example and it's doing the right thing. After teardown is received no more result is dispatched by the fetchExchange.

Wrong! `teardown` comes **BEFORE** the server answer (with possibly new data). The call is still in pending but urql says "I've finished the job! Teardown!" Wrong!

This was working OK before latest graphcache udpates.

> However, I found that the original network request isn't cancelled when it stalls after the initial request succeeds. This is likely down to when you delay your response, i.e. after headers are sent by your API. I can see that this is confusing and I can put back the old cancellation logic.

I think my delay on the backend has nothing to do with urql. I mean, I'm delaying the answer because I'm simulating a slow internet. I'm NOT streaming anything. No header sent before answer, nothing at all, simply like:

1. JS client calls the server

1. the backend wait for 3 seconds, then anwser

1. the client get possibly new data

I don't know what to try anymore to make you understand the problem. I don't know what to debug anymore.

**And I'm afraid of staying here alone without answers from you, 'cause you are the only ones who can understand the problem.**

I hope you won't leave me alone.

**Reply by frederikhors** [reply]
@kitten can you please re-open this issue until a solution is found?

**Reply by frederikhors** [reply]
I don't understand this silence, @kitten.

Did I do something wrong other than pointing out these problems to you?

Or is my code wrong?

Or are you working on some solution?

If I have done something wrong please excuse me.

**Reply by frederikhors** [reply]
@JoviDeCroock sorry for the ping. Can you help me understanding what happened with @kitten here? Please, I am disoriented. I've always had respect for all of you and for the project.

**Reply by JoviDeCroock** [reply]
@frederikhors this is open-source, people can take time off and be unresponsive. Nobody owes you an explanation or help, you aren't paying for them to do so. You are empowered to resolve this yourself, all code is there.

**Reply by frederikhors** [reply]
> @frederikhors this is open-source, people can take time off and be unresponsive. Nobody owes you an explanation or help, you aren't paying for them to do so. You are empowered to resolve this yourself, all code is there.

But I opened an issue because **this is an issue in urql code**, not in mine.

Can you please re-open the issue?

**Reply by kitten** [reply]
This is still all expected behaviour. The original call is aborted, which as discussed prior, is still expected behaviour now to restore past behaviour and prevent duplicate operations.

Furthermore, you're confusing operations with network requests. Not all operations result in a network request. So, if you're tracking active operations and pending operations, operations are only pending if they don't receive a result synchronously, since this easily maps to caching behaviour.

The `offlineExchange` restores requests asynchronously, so it's an exception and relies on rebounding the operation. If it's a cache miss, the request will restart, if not, the original network request will be canceled.

**Reply by frederikhors** [reply]
Thanks for [the reply](https://github.com/urql-graphql/urql/issues/3263#issuecomment-1587057646), @kitten (you can edit it even now, I've read).

1. _I've used `yarn` on CodeSandbox (and the `yarn dedupe` command; I think if it didn't get the right `fetchExchange` that's a problem in itself);_

1. anyway only now with [your answer](https://github.com/urql-graphql/urql/issues/3263#issuecomment-1587057646) I understand that the new `offlineExchange` version `6` instead of version `4` is what you think is the right behaviour; you keep talking about "fixed regression and behavior restored" but I don't think it has been restored;

1. in fact also using the `fetchExchange` of `@urql/core@4.0.10` you say that the call is "**correctly** canceled";

1. how can I now reload that list automatically?
    - What is the `correct` behaviour now?
    
    - With version `4` the `cache-and-network` was working perfectly; now with version `6` the call is canceled.
    
    - **I don't understand what to use now to automatically reload that list when I open the browser and the cached data are rendered.**

1. _I apologize again if I have been a bit "insistent" in the last few days; I have such a congestion of things and this change (which impacts my application a lot) has made me very distressed; I don't think I have been annoying on other occasions because in the end I have always pointed out the right, concrete (and in fact then fixed by you) issues._

The point 4 is my issue.

**Reply by kitten** [reply]
You should be able to use the `requestPolicyExchange`. If that doesn't work that's a separate issue, but I'm not sure if you've tried it yet.

As before, changing the default policy to `cache-and-network` won't always yield the correct. behaviour, and as stated before, is now blocked and overwritten for this exact use-case to protect you from a duplicate request edge case.

As per being insistent, my time is just as congested and limited for long periods of time. However, most importantly, whenever my time is limited, due to there not being any monetary value on urql's development just yet, I can only limit the time I spend on it when I have conflicting gigs ongoing

**Reply by frederikhors** [reply]
I didn't use `requestPolicyExchange` because I don't think it's the way to solve this problem.

Urql is working amazingly well for all the rest in my application.

This is the only issue: **reload of the active (currently visible and with cached data) queries after graphcache starts**.

I don't need `requestPolicyExchange` at all after the start: I'm super-happy with `cache-and-network` as is where needed.

### What brought us to this situation?

Why is it now no longer possible to make this reload automatic like it was with version `4`?

> changing the default policy to cache-and-network won't always yield the correct behaviour

I'm not changing the "default policy", I'm simply using the `cache-and-network` policy for THIS list and this list only.

> is now blocked and overwritten for this exact use-case to protect you from a duplicate request edge case

If you define the reload of an application as an `edge case`, I'm fine with duplicated requests; which, however, until version `4` were managed in a splendid way by the `dedupe` exchange. Is this what do you mean?

### An idea

Is there any chance we can add an option to `offlineExchange`, like: `reloadActiveCachedQueries` or `enableCacheAndNetworkRequestsReload`?

```ts
offlineExchange({
  schema,
  storage,
  keys,
  resolvers,
  updates,
  reloadActiveCachedQueries: true,
}),
```

_For the problem of limited time: I understand you perfectly. I have no problem waiting for an answer (I rolled-back to version `4`): I was just afraid that you would be convinced that the mistake was in my code and that you would not answer me anymore (because I saw that you were answering others). :smile:_

**Reply by frederikhors** [reply]
@kitten, have you thought about my idea?

Can you recommend a way to reload the queries when I reopen the browser page?
